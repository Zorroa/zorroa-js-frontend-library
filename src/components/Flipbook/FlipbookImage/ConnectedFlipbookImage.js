import { connect } from 'react-redux'
import FlipbookImage from './FlipbookImage'

export default connect(state => ({
  fps: state.app.flipbookFps,
  origin: state.auth.origin,
  shouldLoop: state.app.shouldLoop,
  shouldHold: state.app.shouldHold,
}))(FlipbookImage)
