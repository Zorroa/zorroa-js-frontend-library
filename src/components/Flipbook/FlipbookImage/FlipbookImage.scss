.FlipbookImage {
  // Think long and hard before changing the font size, you could introduce
  // excess white space in certain layouts that weren't expecting any text.
  font-size: 0;
  position: relative;

  &,
  &__loading-status {
    display: flex;
    align-items: center;
    align-content: center;
    justify-content: center;
    width: 100%;
    height: 100%;
  }

  &__loading-status {
    top: 0;
    left: 0;
    z-index: 1;
    position: absolute;

    &--no-frame-metadata {
      position: static;
    }
  }
}
