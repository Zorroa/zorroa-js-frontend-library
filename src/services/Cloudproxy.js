import axios from 'axios'
import CloudproxyStats from '../models/CloudproxyStats'
import CloudproxySettings from '../models/CloudproxySettings'
import FilesystemEntry from '../models/FilesystemEntry'

export default class Cloudproxy {
  stats = null
  statInterval = null
  onProgress = null

  constructor(server, onProgress) {
    this.onProgress = onProgress
    if (server) this.initialize(server)
  }

  initialize(server) {
    const url = 'http://' + server + ':8090'
    if (this.cloudproxy && this.cloudproxy.baseURL === url) return
    this.cloudproxy = axios.create({
      baseURL: url,
      headers: { withCredentials: true },
    })
  }

  start() {
    if (!this.cloudproxy) return
    if (this.statInterval) clearInterval(this.statInterval)
    this.getStats(true /* immediate */)
    this.statInterval = setInterval(this.getStats, 1500)
  }

  stop() {
    clearInterval(this.statInterval)
    this.statInterval = null
  }

  getStats = immediate => {
    if (!this.cloudproxy) return
    this.cloudproxy
      .get('/api/v1/stats', { timeout: 1000 })
      .then(response => {
        console.log('CloudproxyStats: ' + JSON.stringify(response.data))
        if (!this.stats)
          setTimeout(_ => this.onProgress(this.stats), immediate ? 1 : 1500)
        this.stats = new CloudproxyStats(response.data)
      })
      .catch(error => {
        console.log('Error getting stats: ' + error)
      })
  }

  import = (archivistUrl, hmacKey, authUser, paths) => {
    const settings = new CloudproxySettings({
      archivistUrl,
      hmacKey,
      authUser,
      paths,
      startNow: true,
      schedule: null,
      threads: 1,
      pipelineId: -1,
    })
    console.log('Starting cloudproxy: ' + JSON.stringify(settings))
    this.cloudproxy
      .put('/api/v1/settings', settings)
      .then(response => {
        console.log('Started cloudproxy: ' + JSON.stringify(response.data))
      })
      .catch(error => {
        console.log('Cloudproxy error: ' + error)
      })
  }

  cancel = () => {
    if (!this.stats || !this.stats.active) return
    console.log('Canceling cloudproxy')
    this.cloudproxy
      .delete('/api/v1/import')
      .then(response => {
        console.log('Deleted ' + JSON.stringify(response.data))
      })
      .catch(error => {
        console.log('Error canceling cloudproxy: ' + error)
      })
  }

  listDirectory = (path, closure) => {
    this.cloudproxy
      .put('/api/v1/files/_path', { path })
      .then(response => {
        console.log(
          'Got files at ' + path + ': ' + JSON.stringify(response.data),
        )
        const files = response.data.map(f => new FilesystemEntry(f))
        closure(files)
      })
      .catch(console.log('Error getting files at path: ' + path))
  }
}
