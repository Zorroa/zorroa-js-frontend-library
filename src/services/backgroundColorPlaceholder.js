export default function getBackgroundPlaceholder(colors, randomSeed) {
  const defaultColors = ['#73b61c', '#89c366', '#51af5f', '#55c39e', '#b4a50b']
  const placeholderColors = Array.isArray(colors) ? colors : defaultColors
  const positions = [
    'top left',
    'top center',
    'top right',
    'center left',
    'center center',
    'center right',
    'bottom right',
    'bottom center',
    'bottom left',
  ]
  const backgroundPlaceholder = placeholderColors
    .slice(0, positions.length)
    .map((color, index, array) => {
      const changeFactor = randomSeed % 2 === 0 ? -5 : 5
      const size = 80 + changeFactor / array.length * index
      const newGradient = `radial-gradient(circle at ${
        positions[index]
      }, ${color}, rgba(255,0,0,0) ${size}%)`
      return newGradient
    })
    .join(', ')
  return `${backgroundPlaceholder}`
}
