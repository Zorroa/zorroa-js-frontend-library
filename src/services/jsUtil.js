export function unCamelCase (str) {
  return str && str
    // insert a space between lower & upper
    .replace(/([a-z])([A-Z])/g, '$1 $2')
    // space before last upper in a sequence followed by lower
    .replace(/\b([A-Z]+)([A-Z])([a-z])/, '$1 $2$3')
    // uppercase the first character
    .replace(/^./, function (str) { return str.toUpperCase() })
}

export function formatDuration (seconds, fps) {
  const date = new Date(seconds * 1000)
  const hh = date.getUTCHours()
  const mm = date.getUTCMinutes()
  const ss = pad(date.getUTCSeconds())
  if (hh) {
    return `${hh}:${pad(mm)}:${ss}`
  } else if (fps) {
    const ff = pad(Math.round(fps * date.getUTCMilliseconds() / 1000))
    return `${mm}:${ss}:${ff}`
  }
  return `${mm}:${ss}`
}

// SPE's time code format is hours:minutes:seconds:frames
const timeRE = /(\d+):(\d+):(\d+):(\d+)/
//
export function parseTimecodeMS (timecodeStr) {
  const matches = timeRE.exec(timecodeStr)
  if (matches.length !== 5) return 0
  const numbers = matches.slice(1).map(x => parseInt(x, 10))
  var time = 0
  time += numbers[0] * 3600 * 1000 // millis per hour
  time += numbers[1] * 60 * 1000 // millis per minute
  time += numbers[2] * 1 * 1000 // millis per second
  time += numbers[3] * (1 / 23.98) * 1000 // millis per frame (sony data all has 23.98 frames/sec)
  return time
}

export function remap (x, min1, max1, min2, max2) {
  return min2 + (max2 - min2) * ((x - min1) / (max1 - min1))
}

export function clamp (v, min, max) {
  return Math.max(min, Math.min(max, v))
}

function pad (string, digits) {
  return ('0' + string).slice(-(digits || 2))
}

export function parseFormattedFloat (obj) {
  if (!obj) return
  if (typeof obj === 'number') return obj
  if (typeof obj === 'string') {
    return parseFloat(obj.replace(',', ''))
  }
}

export function humanFileSize (size) {
  const i = Math.floor(Math.log(size) / Math.log(1024))
  return (size / Math.pow(1024, i)).toFixed(2) * 1 + ' ' + ['B', 'kB', 'MB', 'GB', 'TB'][i]
}

// Returns a set ids that either contains the currently selected ids
// or the isolated id depending on whether the isolated id is in the
// selected set.
export function isolateSelectId (id, selectedIds) {
  if (selectedIds && selectedIds.size && selectedIds.has(id)) {
    return selectedIds
  }
  return new Set([id])
}

// Add all sibling assets in allAssets to assetIds.
// WARNING: requires multiple passes over allAssets.
export function addSiblings (assetIds, allAssets) {
  const parentIds = new Set()
  assetIds.forEach(id => {
    const index = allAssets.findIndex(asset => (asset.id === id))
    const parentId = allAssets[index].parentId()
    if (parentId) parentIds.add(parentId)
  })
  allAssets.forEach(asset => {
    const parentId = asset.parentId()
    if (parentId && parentIds.has(parentId)) assetIds.add(asset.id)
  })
}

export function equalSets (as, bs) {
  if (as.size !== bs.size) return false
  for (var a of as) if (!bs.has(a)) return false
  return true
}

/* ----------------------------------------------------------------------
Execute a sequence of promises, with a max limit
on the number of promises in flight at once.

'data' is an array containing the data needed to make each promise
  each element of 'data' is passed to 'mkPromiseFn'
'mkPromiseFn' is a callback function that generates a promise
  called once per item in 'data', passed the single datum, takes only that 1 param
'optQueueSize' is an OPTIONAL number specifying the max queue length
  negative optQueueSize == #data/n (percentage), e.g. -1 for no limit, -2 for n/2, etc...
  default optQueueSize == sqrt(data.length)
'optProgressFn' is an OPTIONAL callback that is called once for every completed promise
and passed the number of completed items

Returns a promise that resolves when all promises in the queue resolve,
or rejects when the first promise in the queue rejects,
just like Promise.all()

Resolves with an array of the results, in the same order as 'data'
Rejects with the first error that occurs
*/
export function makePromiseQueue (data, mkPromiseFn, optQueueSize, optProgressFn) {
  return new Promise((resolve, reject) => {
    const n = data.length
    optQueueSize = optQueueSize || Math.round(Math.sqrt(n))
    if (optQueueSize < 0) optQueueSize = n / Math.abs(optQueueSize)
    optQueueSize = Math.max(1, Math.min(n, optQueueSize))

    let dataResults = []
    let nextDataToProcess = 0
    let totalFinished = 0

    let start = (idx) => {
      nextDataToProcess++
      mkPromiseFn(data[idx])
      .then((results) => {
        totalFinished++
        dataResults[idx] = results
        if (totalFinished >= n) resolve(dataResults)
        else optProgressFn && optProgressFn(totalFinished, n)
        if (nextDataToProcess < n) start(nextDataToProcess)
      })
      .catch(err => reject(err))
    }

    for (var i = 0; i < optQueueSize; i++) start(i)
  })
}
