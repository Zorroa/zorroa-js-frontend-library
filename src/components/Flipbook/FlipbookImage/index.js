import React, { Component, PropTypes } from 'react'
import ConnectedFlipbookImage from './ConnectedFlipbookImage'

import api from '../../../api'
import { PubSub } from '../../../services/jsUtil'
import Asset from '../../../models/Asset'

export default class FlipbookImageContainer extends Component {
  static propTypes = {
    clipParentId: PropTypes.string.isRequired,
    defaultFrame: PropTypes.instanceOf(Asset),
    autoPlay: PropTypes.bool,
    status: PropTypes.instanceOf(PubSub),
    shuttler: PropTypes.instanceOf(PubSub),
  }

  componentDidMount() {
    this.getFlipbook()
  }

  componentDidReceiveProps(prevProps) {
    if (prevProps.clipParentId !== this.props.clipParentId) {
      this.getFlipbook()
    }
  }

  getFlipbook() {
    this.setState({
      frames: [],
      isLoading: true,
      isError: false,
    })

    api.flipbook.get(this.props.clipParentId).then(
      frames => {
        this.setState({
          frames,
          isLoading: false,
          isError: false,
        })
      },
      () => {
        this.setState({
          frames: [],
          isError: true,
          isLoading: false,
        })
      },
    )
  }

  render() {
    return <ConnectedFlipbookImage {...this.state} {...this.props} />
  }
}
