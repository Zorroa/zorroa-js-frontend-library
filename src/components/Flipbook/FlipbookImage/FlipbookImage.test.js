/* eslint-env jest */

import React from 'react'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-15'
import FlipbookImage from './FlipbookImage'
import Asset from '../../../models/Asset'
import { PubSub } from '../../../services/jsUtil'
import renderer from 'react-test-renderer'

const origin = 'http://localhost'
configure({ adapter: new Adapter() })

function generateFrameAsset(id, number) {
  return new Asset({
    id: id || `${Math.round(Math.random() * 100000)}`,
    score: NaN,
    document: {
      media: {
        clip: {
          type: 'flipbook',
          start:
            typeof number === 'number' ? number : Math.round(Math.random * 10),
          stop:
            typeof number === 'number' ? number : Math.round(Math.random * 10),
        },
      },
    },
  })
}
describe('<FlipbookImage />', () => {
  describe('animationHoldStateUpdate()', () => {
    const frames = [
      generateFrameAsset('a', 2),
      generateFrameAsset('b', 5),
      generateFrameAsset('c', 6),
    ]
    const componentInstance = shallow(
      <FlipbookImage
        shouldLoop={true}
        shouldHold={true}
        origin={origin}
        frames={frames}
      />,
    )
    componentInstance.instance().animationHoldStateUpdate()
    expect(componentInstance.state('activeFrame')).toBe(frames[0])
  })

  describe('getClosestFrameByPercent()', () => {
    const frames = [
      generateFrameAsset('a', 1),
      generateFrameAsset('b', 10),
      generateFrameAsset('c', 100),
    ]
    const componentInstance = shallow(
      <FlipbookImage
        shouldLoop={true}
        shouldHold={true}
        origin={origin}
        frames={frames}
      />,
    )
    const zeroPercent = componentInstance.instance().getClosestFrameByPercent(0)
    const fivePercent = componentInstance
      .instance()
      .getClosestFrameByPercent(0.05)
    const fifteenPercent = componentInstance
      .instance()
      .getClosestFrameByPercent(0.15)
    const fiftyPercent = componentInstance
      .instance()
      .getClosestFrameByPercent(0.5)
    const seventyFivePercent = componentInstance
      .instance()
      .getClosestFrameByPercent(0.75)
    const oneHundredPercent = componentInstance
      .instance()
      .getClosestFrameByPercent(1)
    expect(zeroPercent).toBe(frames[0])
    expect(fivePercent).toBe(frames[0])
    expect(fifteenPercent).toBe(frames[1])
    expect(fiftyPercent).toBe(frames[1])
    expect(seventyFivePercent).toBe(frames[2])
    expect(oneHundredPercent).toBe(frames[2])
  })

  describe('notifySubscribeesOfNewFrame()', () => {
    it('Should publish the elapsed amount', () => {
      let elapsedPercent
      const status = new PubSub()
      status.on('elapsedPercent', percent => {
        elapsedPercent = percent
      })
      const frames = [
        generateFrameAsset(undefined, 1),
        generateFrameAsset(undefined, 2),
        generateFrameAsset(undefined, 3),
      ]
      const componentInstance = shallow(
        <FlipbookImage status={status} origin={origin} frames={frames} />,
      )

      componentInstance.instance().setActiveFrame(frames[0])
      componentInstance.instance().notifySubscribeesOfNewFrame()
      expect(elapsedPercent).toBe(0)
      componentInstance.instance().setActiveFrame(frames[1])
      componentInstance.instance().notifySubscribeesOfNewFrame()
      expect(elapsedPercent).toBe(0.5)
      componentInstance.instance().setActiveFrame(frames[2])
      componentInstance.instance().notifySubscribeesOfNewFrame()
      expect(elapsedPercent).toBe(1)
    })
  })

  describe('updateDefaultFrameElapsedPosition()', () => {
    describe('When a default frame is defined', () => {
      it('Should publish the elapsed amount', () => {
        let elapsedPercent
        const status = new PubSub()
        status.on('elapsedPercent', percent => {
          elapsedPercent = percent
        })
        const frames = [
          generateFrameAsset(undefined, 1),
          generateFrameAsset(undefined, 2),
          generateFrameAsset(undefined, 3),
        ]
        const componentInstance = shallow(
          <FlipbookImage
            status={status}
            defaultFrame={frames[1]}
            origin={origin}
            frames={frames}
          />,
        )

        componentInstance.instance().updateDefaultFrameElapsedPosition()
        expect(elapsedPercent).toBe(0.5)
      })
    })

    describe('When a default frame is not defined', () => {
      it('Should publish the elapsed amount as not started', () => {
        let elapsedPercent
        const status = new PubSub()
        status.on('elapsedPercent', percent => {
          elapsedPercent = percent
        })
        const frames = [
          generateFrameAsset(undefined, 1),
          generateFrameAsset(undefined, 2),
        ]
        const componentInstance = shallow(
          <FlipbookImage status={status} origin={origin} frames={frames} />,
        )

        componentInstance.instance().updateDefaultFrameElapsedPosition()
        expect(elapsedPercent).toBe(0)
      })
    })
  })

  describe('When in "hold frames" mode', () => {
    describe('When the frame numbers have dropped frames', () => {
      it('Should publish the elapsed amount in terms of total frame count', () => {
        let elapsedPercent
        const status = new PubSub()
        status.on('elapsedPercent', percent => {
          elapsedPercent = percent
        })
        const frames = [
          generateFrameAsset(undefined, 1),
          generateFrameAsset(undefined, 5),
        ]
        const componentInstance = shallow(
          <FlipbookImage
            status={status}
            shouldHold={true}
            origin={origin}
            frames={frames}
          />,
        )

        componentInstance.instance().animationHoldStateUpdate()
        expect(elapsedPercent).toBe(0)
        componentInstance.instance().animationHoldStateUpdate()
        expect(elapsedPercent).toBe(0.25)
        componentInstance.instance().animationHoldStateUpdate()
        expect(elapsedPercent).toBe(0.5)
        componentInstance.instance().animationHoldStateUpdate()
        expect(elapsedPercent).toBe(0.75)
        componentInstance.instance().animationHoldStateUpdate()
        expect(elapsedPercent).toBe(1)
      })
    })

    describe('When the frame numbers have sequential frames', () => {
      it('Should publish the elapsed amount in terms of total frame count', () => {
        let elapsedPercent
        const status = new PubSub()
        status.on('elapsedPercent', percent => {
          elapsedPercent = percent
        })
        const frames = [
          generateFrameAsset(undefined, 0),
          generateFrameAsset(undefined, 1),
          generateFrameAsset(undefined, 2),
        ]
        const componentInstance = shallow(
          <FlipbookImage
            status={status}
            shouldHold={true}
            origin={origin}
            frames={frames}
          />,
        )

        componentInstance.instance().animationHoldStateUpdate()
        expect(elapsedPercent).toBe(0)
        componentInstance.instance().animationHoldStateUpdate()
        expect(elapsedPercent).toBe(0.5)
        componentInstance.instance().animationHoldStateUpdate()
        expect(elapsedPercent).toBe(1)
      })
    })
  })

  describe('animationStateUpdate()', () => {
    describe('When looping is enabled', () => {
      it('Should play all frames in a loop without stopping at the end', () => {
        const frames = [generateFrameAsset('1', 1), generateFrameAsset('2', 2)]
        const componentInstance = shallow(
          <FlipbookImage shouldLoop origin={origin} frames={frames} />,
        )

        componentInstance.instance().animationStateUpdate()
        expect(componentInstance.state('activeFrame')).toBe(frames[0])
        componentInstance.instance().animationStateUpdate()
        expect(componentInstance.state('activeFrame')).toBe(frames[1])
        componentInstance.instance().animationStateUpdate()
        expect(componentInstance.state('activeFrame')).toBe(frames[0])
      })
    })

    describe('When looping is disabled', () => {
      it('Should play all frames and then stop', () => {
        const frames = [generateFrameAsset('1', 1), generateFrameAsset('2', 2)]
        const componentInstance = shallow(
          <FlipbookImage shouldLoop={false} origin={origin} frames={frames} />,
        )

        componentInstance.instance().animationStateUpdate()
        expect(componentInstance.state('activeFrame')).toBe(frames[0])
        componentInstance.instance().animationStateUpdate()
        expect(componentInstance.state('activeFrame')).toBe(frames[1])
        componentInstance.instance().animationStateUpdate()
        expect(componentInstance.state('activeFrame')).toBe(frames[1])
        expect(componentInstance.instance().isPlaying()).toBe(false)
      })
    })
  })

  describe('isLoaded()', () => {
    describe('When frames have not loaded', () => {
      const frames = [
        generateFrameAsset(undefined, 1),
        generateFrameAsset(undefined, 2),
      ]
      const componentInstance = shallow(
        <FlipbookImage origin={origin} frames={frames} />,
      )

      expect(componentInstance.instance().isLoaded()).toBe(false)
      componentInstance.instance().onFrameLoad()
      expect(componentInstance.instance().isLoaded()).toBe(false)
    })

    describe('When all frames have loaded', () => {
      const frames = [
        generateFrameAsset(undefined, 1),
        generateFrameAsset(undefined, 2),
      ]
      const componentInstance = shallow(
        <FlipbookImage origin={origin} frames={frames} />,
      )

      frames.forEach(componentInstance.instance().onFrameLoad)
      expect(componentInstance.instance().isLoaded()).toBe(true)
    })
  })

  describe('When no frames are loaded', () => {
    it('It should render', () => {
      const tree = renderer
        .create(<FlipbookImage origin={origin} autoplay />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('scrubByPercent()', () => {
    describe('When the frames are not being held', () => {
      const frames = [
        generateFrameAsset(undefined, 1),
        generateFrameAsset(undefined, 2),
        generateFrameAsset(undefined, 3),
        generateFrameAsset(undefined, 4),
      ]
      const componentInstance = shallow(
        <FlipbookImage origin={origin} frames={frames} />,
      )

      it('Should navigate to the proper frame', () => {
        componentInstance.instance().scrubByPercent(0.8)
        expect(componentInstance.state('activeFrame')).toEqual(frames[3])
        componentInstance.instance().scrubByPercent(0.1)
        expect(componentInstance.state('activeFrame')).toBe(frames[0])
        componentInstance.instance().scrubByPercent(0.5)
        expect(componentInstance.state('activeFrame')).toBe(frames[2])
      })
    })

    describe('When the frames are being held', () => {
      const frames = [
        generateFrameAsset(undefined, 0),
        generateFrameAsset(undefined, 5),
        generateFrameAsset(undefined, 10),
        generateFrameAsset(undefined, 20),
      ]
      const componentInstance = shallow(
        <FlipbookImage origin={origin} shouldHold={true} frames={frames} />,
      )

      it('Should navigate to the proper frame', () => {
        componentInstance.instance().scrubByPercent(0.01)
        expect(componentInstance.state('activeFrame')).toEqual(frames[0])
        componentInstance.instance().scrubByPercent(0.2)
        expect(componentInstance.state('activeFrame')).toEqual(frames[1])
        componentInstance.instance().scrubByPercent(0.6)
        expect(componentInstance.state('activeFrame')).toBe(frames[2])
        componentInstance.instance().scrubByPercent(0.85)
        expect(componentInstance.state('activeFrame')).toBe(frames[3])
      })
    })
  })

  describe('frameForward()', () => {
    const frames = [
      generateFrameAsset(undefined, 1),
      generateFrameAsset(undefined, 2),
      generateFrameAsset(undefined, 3),
    ]
    const componentInstance = shallow(
      <FlipbookImage origin={origin} frames={frames} />,
    )

    it('Should pause the animation', () => {
      componentInstance.instance().start()

      // Ensure the animation is unpaused first
      expect(componentInstance.instance().isPaused()).toBe(false)

      // Then go forward a frame
      componentInstance.instance().frameForward()

      // Then expect the animation is paused
      expect(componentInstance.instance().isPaused()).toBe(true)
    })

    it('Should go forward one frame', () => {
      componentInstance.instance().setActiveFrame(frames[0])
      componentInstance.instance().frameForward()
      expect(componentInstance.state('activeFrame')).toBe(frames[1])
      componentInstance.instance().frameForward()
      expect(componentInstance.state('activeFrame')).toBe(frames[2])
      componentInstance.instance().frameForward()
      expect(componentInstance.state('activeFrame')).toBe(frames[2])
    })
  })

  describe('frameBack()', () => {
    const frames = [
      generateFrameAsset(undefined, 1),
      generateFrameAsset(undefined, 2),
      generateFrameAsset(undefined, 3),
    ]
    const componentInstance = shallow(
      <FlipbookImage origin={origin} frames={frames} />,
    )

    it('Should pause the animation', () => {
      componentInstance.instance().start()

      // Ensure the animation is unpaused first
      expect(componentInstance.instance().isPaused()).toBe(false)

      // Then go forward a frame
      componentInstance.instance().frameBack()

      // Then expect the animation is paused
      expect(componentInstance.instance().isPaused()).toBe(true)
    })

    it('Should go backward one frame', () => {
      componentInstance.instance().setActiveFrame(frames[2])
      componentInstance.instance().frameBack()
      expect(componentInstance.state('activeFrame')).toBe(frames[1])
      componentInstance.instance().frameBack()
      expect(componentInstance.state('activeFrame')).toBe(frames[0])
      componentInstance.instance().frameBack()
      expect(componentInstance.state('activeFrame')).toBe(frames[0])
    })
  })

  describe('fastForward()', () => {
    const frames = [
      generateFrameAsset(undefined, 1),
      generateFrameAsset(undefined, 2),
      generateFrameAsset(undefined, 3),
    ]
    const componentInstance = shallow(
      <FlipbookImage origin={origin} frames={frames} />,
    )

    it('Should go to the last frame', () => {
      componentInstance.instance().setActiveFrame(frames[1])
      expect(componentInstance.state('activeFrame')).toBe(frames[1])

      componentInstance.instance().fastForward()
      expect(componentInstance.state('activeFrame')).toBe(frames[2])
    })

    it('Should be paused', () => {
      componentInstance.instance().start()
      componentInstance.instance().setActiveFrame(frames[1])
      expect(componentInstance.state('activeFrame')).toBe(frames[1])

      componentInstance.instance().fastForward()
      expect(componentInstance.instance().isPlaying()).toBe(false)
    })
  })

  describe('startOrStop()', () => {
    describe('When the animation is at the end', () => {
      const frames = [
        generateFrameAsset(undefined, 1),
        generateFrameAsset(undefined, 2),
        generateFrameAsset(undefined, 3),
      ]
      const componentInstance = shallow(
        <FlipbookImage origin={origin} frames={frames} />,
      )

      it('Should start playing the first frame', () => {
        componentInstance.instance().setActiveFrame(frames[2])
        componentInstance.instance().startOrStop()
        expect(componentInstance.state('activeFrame')).toBe(frames[0])
        expect(componentInstance.instance().isPlaying()).toBe(true)
      })
    })

    describe('When the animation is paused', () => {
      const frames = [generateFrameAsset(undefined, 1)]
      const componentInstance = shallow(
        <FlipbookImage origin={origin} frames={frames} />,
      )

      it('Should become unpaused', () => {
        componentInstance.instance().pause()
        expect(componentInstance.instance().isPaused()).toBe(true)
        componentInstance.instance().startOrStop()
        expect(componentInstance.instance().isPaused()).toBe(false)
      })
    })

    describe('When the animation is playing', () => {
      const frames = [generateFrameAsset(undefined, 1)]
      const componentInstance = shallow(
        <FlipbookImage origin={origin} frames={frames} />,
      )

      it('Should become paused', () => {
        componentInstance.instance().start()
        componentInstance.instance().startOrStop()
        expect(componentInstance.instance().isPaused()).toBe(true)
      })
    })

    describe('When the animation is loaded and not playing', () => {
      const frames = [generateFrameAsset(undefined, 1)]
      const componentInstance = shallow(
        <FlipbookImage origin={origin} frames={frames} />,
      )

      it('Should start playing', () => {
        frames.forEach(componentInstance.instance().onFrameLoad)
        componentInstance.instance().startOrStop()
        expect(componentInstance.instance().isPlaying()).toBe(true)
      })
    })
  })

  describe('getFormattedLoadingPercentage()', () => {
    const frames = [
      generateFrameAsset(undefined, 1),
      generateFrameAsset(undefined, 2),
      generateFrameAsset(undefined, 3),
    ]
    const componentInstance = shallow(
      <FlipbookImage origin={origin} frames={frames} />,
    )

    it('Should display percentages as fraction of 100', () => {
      expect(componentInstance.instance().getFormattedLoadingPercentage()).toBe(
        0,
      )
      componentInstance.instance().onFrameLoad()
      expect(componentInstance.instance().getFormattedLoadingPercentage()).toBe(
        33,
      )
      componentInstance.instance().onFrameLoad()
      expect(componentInstance.instance().getFormattedLoadingPercentage()).toBe(
        67,
      )
      componentInstance.instance().onFrameLoad()
      expect(componentInstance.instance().getFormattedLoadingPercentage()).toBe(
        100,
      )
    })
  })

  describe('isAtEndOfAnimation()', () => {
    describe('When the animation is in the middle', () => {
      const frames = [
        generateFrameAsset(undefined, 1),
        generateFrameAsset(undefined, 2),
      ]
      const componentInstance = shallow(
        <FlipbookImage origin={origin} frames={frames} />,
      )

      it('Should be false', () => {
        componentInstance.instance().setActiveFrame(frames[0])
        expect(componentInstance.instance().isAtEndOfAnimation()).toBe(false)
      })

      describe('When the animation is at the end', () => {
        const frames = [
          generateFrameAsset(undefined, 1),
          generateFrameAsset(undefined, 2),
        ]
        const componentInstance = shallow(
          <FlipbookImage origin={origin} frames={frames} />,
        )

        it('Should be true', () => {
          componentInstance.instance().setActiveFrame(frames[1])
          expect(componentInstance.instance().isAtEndOfAnimation()).toBe(true)
        })
      })
    })
  })

  describe('rewind()', () => {
    describe('When the animation is playing', () => {
      const frames = [
        generateFrameAsset(undefined, 1),
        generateFrameAsset(undefined, 2),
        generateFrameAsset(undefined, 3),
      ]
      const componentInstance = shallow(
        <FlipbookImage origin={origin} frames={frames} />,
      )

      it('Should go to the first frame', () => {
        componentInstance.instance().start()
        componentInstance.instance().setActiveFrame(frames[2])

        // Ensure the current frame is not already the first frame
        expect(componentInstance.state('activeFrame')).toBe(frames[2])

        componentInstance.instance().rewind()
        expect(componentInstance.state('activeFrame')).toBe(frames[0])
        componentInstance.instance().stop()
      })

      it('Should keep playing', () => {
        componentInstance.instance().start()
        componentInstance.instance().rewind()
        expect(componentInstance.instance().isPlaying()).toBe(true)
      })
    })

    describe('When the animation is not playing', () => {
      const frames = [
        generateFrameAsset(undefined, 1),
        generateFrameAsset(undefined, 2),
        generateFrameAsset(undefined, 3),
      ]
      const componentInstance = shallow(
        <FlipbookImage origin={origin} frames={frames} />,
      )

      it('Should go to the first frame', () => {
        componentInstance.instance().stop()
        componentInstance.instance().setActiveFrame(frames[2])

        // Ensure the current frame is not already the first frame
        expect(componentInstance.state('activeFrame')).toBe(frames[2])

        componentInstance.instance().rewind()
        expect(componentInstance.state('activeFrame')).toBe(frames[0])
      })

      it('Should stay paused', () => {
        componentInstance.instance().rewind()
        expect(componentInstance.instance().isPlaying()).toBe(false)
      })
    })
  })

  describe('shouldFrameBeVisible()', () => {
    describe('When there is no default frame', () => {
      const frames = [
        generateFrameAsset(undefined, 1),
        generateFrameAsset(undefined, 2),
      ]
      const componentInstance = shallow(
        <FlipbookImage origin={origin} frames={frames} />,
      )

      describe('When the images are still loading', () => {
        it('Should never return a visible frame', () => {
          expect(
            componentInstance.instance().shouldFrameBeVisible(frames[0]),
          ).toBe(false)
          expect(
            componentInstance.instance().shouldFrameBeVisible(frames[1]),
          ).toBe(false)
        })
      })

      describe('When the images are loaded', () => {
        it('Should return true for the active frame', () => {
          componentInstance.instance().onFrameLoad()
          componentInstance.instance().onFrameLoad()
          expect(
            componentInstance.instance().shouldFrameBeVisible(frames[0]),
          ).toBe(true)
          expect(
            componentInstance.instance().shouldFrameBeVisible(frames[1]),
          ).toBe(false)
        })
      })
    })

    describe('When there is a default frame', () => {
      const frames = [
        generateFrameAsset('100', 1),
        generateFrameAsset('101', 2),
      ]
      const defaultFrame = frames[1]
      const componentInstance = shallow(
        <FlipbookImage
          defaultFrame={defaultFrame}
          origin={origin}
          frames={frames}
        />,
      )

      it('Should return true only for the default frame', () => {
        expect(
          componentInstance.instance().shouldFrameBeVisible(frames[0]),
        ).toBe(false)
        expect(
          componentInstance.instance().shouldFrameBeVisible(defaultFrame),
        ).toBe(true)
      })
    })
  })
})
