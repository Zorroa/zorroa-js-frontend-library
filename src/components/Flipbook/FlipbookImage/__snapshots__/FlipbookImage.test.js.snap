// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlipbookImage /> When no frames are loaded It should render 1`] = `
<div>
  <div
    className="FlipbookImage"
  >
    <div
      className="FlipbookImage__loading-status FlipbookImage__loading-status--no-frame-metadata"
      style={
        Object {
          "height": undefined,
          "width": undefined,
        }
      }
    >
      <svg
        className="ProgressCircle"
        height={120}
        viewBox="0 0 120 120"
        width={120}
      >
        <circle
          className="background"
          cx={60}
          cy={60}
          r={57.5}
          strokeWidth="5px"
        />
        <circle
          className="foreground"
          cx={60}
          cy={60}
          r={57.5}
          strokeWidth="5px"
          style={
            Object {
              "strokeDasharray": 361.28315516282623,
              "strokeDashoffset": 361.28315516282623,
            }
          }
        />
        <text
          className="text"
          dy=".4em"
          textAnchor="middle"
          x={60}
          y={60}
        >
          0%
        </text>
      </svg>
    </div>
  </div>
</div>
`;
